Make sure embeddings don't fail

Til7701 opened the pull request:
- When documents are ready to be embedded, they are now added to an `EmbeddingBacklog`
- The `EmbeddingTaskExecutor` takes document from the `EmbeddingBacklog` and tries to embed them.
- If this fails, it tries again after some time.
- If it fails too many times in a row it waits longer and eventually stops the thread. This might be changed to stopping the application as a whole.

I also tested other methods to prevent hitting the rate limit. When we are using an embedding model with a larger context window, we may increase the batch size for requests and not the length of the actual documents. Thus, the chunks of test are not too large, and we still get a performance boost due to fewer calls to the model.

A database version of the `EmbeddingBacklog` may be implemented to avoid loosing data upon restart of the application.

Til7701 committed:
WIP Make sure embeddings don't fail with rate limits

- When documents are ready to be embedded, they are now added to an EmbeddingBacklog
- The EmbeddingTaskExecutor takes document from the EmbeddingBacklog and tries to embed them.
- If this fails, it tries again after some time.

I also tested other methods to prevent hitting the rate limit. When we are using an embedding model with a larger context window, we may increase the batch size for requests and not the length of the actual documents. Thus, the chunks of test are not too large, and we still get a performance boost due to fewer calls to the model.

TODO:

- [ ] differentiate between exceptions in the EmbeddingTaskExecutor
- [ ] decide between batch size and total token count configuration properties
- [ ] maybe implement a database version of the EmbeddingBacklog
- [ ] make it possible to wait for task completion

Til7701 committed:
Merge remote-tracking branch 'origin/dev' into 118-handle-immediate-faliure-of-embedding-due-to-rate-limit

# Conflicts:
#	src/main/java/de/uol/pgdoener/civicsage/business/embedding/EmbeddingService.java
#	src/main/resources/application.yaml

Til7701 committed:
WIP

Til7701 committed:
Set embedded additional property when listing sources

Til7701 committed:
Handling of errors and correct processing of tasks if errors occur

Til7701 committed:
Fix loop condition

JayPi4c commented on src/main/java/de/uol/pgdoener/civicsage/business/embedding/EmbeddingBacklog.java:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * An EmbeddingBacklog is a queue-like structure that holds embedding tasks.
+ * Implementations of this interface should preserve the order of tasks
```
Why should they preserve the order? In my mind, a selective Backlog would be perfectly fine as well. It could be used as a priority list, ordering what resources need to be indexed first.

Til7701 replied:
I wanted to make sure, that we don't have to deal with starvation. Furthermore, it just seems fair that the first request to index something also gets embedded first. Thus users don't have to wait indefinitely.

I could modify the comment to better reflect those requirements.

JayPi4c commented on src/main/java/de/uol/pgdoener/civicsage/business/embedding/EmbeddingService.java:
```diff
@@ -52,4 +51,19 @@ public void delete(UUID sourceId) {
+            cacheNames = CachingConfig.SEARCH_CACHE_NAME,
+            allEntries = true
+    )
+    public void clearCache() {
```
How silly, that the whole logic is done using a single annotation.

JayPi4c reviewed (commented):
I like the current implementation and see the benefits. I see many problems, that are mitigated by this approach - including the rate limit - however, I don't like the general reasoning to primarily target the rate limit. Personally, I think the rate limit is not a concern of the backend but of the admin who is configuring the application. So if we hit the rate limit, we can give a log and move on. 
But again, this approach does handle more than simply the rate limit which is why I want this to be merged. As of now, however, I can't approve it, as I am not home to test the functionality in depth. (How does it handles loosing network connection or a non responsive llm?). I'll come back to this later this week.

Til7701 commented:
What is your opinion on the mechanism to handle too many failures?

Til7701 committed:
Fix peek in InMemoryEmbeddingBacklog

Previously, the peek function also removed the task. Now the task is added back to the backlog after a task was returned.

There does not seem to be a method in BlockingDeque that blocks until an element becomes available and does NOT remove that element.

Til7701 committed:
Update JavaDoc of the EmbeddingBacklog interface to better reflect requirements for implementations

Til7701 committed:
Remove EmbeddingTasks from EmbeddingBacklog if the source is deleted

When a source is deleted, the EmbeddingBacklog is tasked to remove the associated task, if it exists.

There is still a race condition. If the task was removed while the EmbeddingTaskExecutor already has the task and is not done yet, the source could show up in the VectorStore. I am not sure yet, what to do about that.

Til7701 commented:
TODO

- [x] Handle deletion of source while embedding
- [x] Stagger bootstrap indexing

JayPi4c commented:
> What is your opinion on the mechanism to handle too many failures?

What exactly are you referring to? The implementation as a whole or the `handleException` method in particular? On first glance, both seem ok.

Til7701 commented:
I am referring to the `handleException` method. So it's fine :)

Til7701 committed:
WIP Adding priorities to the EmbeddingBacklog

TODO:

 - Race condition in the InMemoryEmbeddingBacklog

Til7701 committed:
Move EmbeddingBacklog related classes to backlog package

Til7701 committed:
Reimplement InMemoryEmbeddingBacklog with Locks and Conditions; add specific equals and hashCode implementations for EmbeddingTask

Til7701 committed:
Merge remote-tracking branch 'origin/dev' into 118-handle-immediate-faliure-of-embedding-due-to-rate-limit

# Conflicts:
#	src/main/java/de/uol/pgdoener/civicsage/business/embedding/EmbeddingService.java

Til7701 committed:
Make defer atomic in InMemoryEmbeddingBacklog

Til7701 committed:
Wait for completion of task if it was in the backlog and currently processed

Til7701 committed:
Downgrade spring ai bom to 1.0.0 due to swagger ui not working

Til7701 committed:
Fixing return value of remove method
